src/main/java/software/amazon/awssdk/crt/http/HttpConnection.java

```diff
@@ -0,0 +1,334 @@
+            this.request = request;
+            this.responseHandler = responseHandler;
+        }
+
+        /**
+         * Called from Native when new Http Headers have been received.
+         * Note that this function may be called multiple times as HTTP headers are received.
+         * @param responseStatusCode The HTTP Response Status Code
+         * @param nextHeaders The headers received in the latest IO event.
+         */
+        void onHeaders(int responseStatusCode, Header[] nextHeaders){
+            responseHandler.receiveHeaders(responseStatusCode, nextHeaders);
+        }
+
+        /**
+         * Called from Native once all HTTP Headers are processed.
+         * @param hasBody True if the HTTP Response had a Body, false otherwise.
+         */
+        void onHeadersDone(boolean hasBody){
+            responseHandler.receiveHeadersDone(hasBody);
+        }
+
+        /**
+         * Called when new Body bytes have been received.
+         * Note that this function may be called multiple times as headers are received.
+         * @param nextBodyBytes The HTTP Body Bytes received in the last IO Event.
+         * @param currWindowSize The current IO Window Size
+         * @return The new IO Window Size
+         */
+        int onResponseBody(byte[] nextBodyBytes, int currWindowSize){
+            responseHandler.receiveResponseBody(nextBodyBytes);
+            return currWindowSize;
+        }
+
+
+        /**
+         * Called from Native when the ResponseBody has completed.
+         * If successful, errorCode will be 0, otherwise the error String can be retrieved by calling
+         * CRT.awsErrorString(errorCode);
+         * @param errorCode The AWS CommonRuntime errorCode of the Response. Will be 0 if no error.
+         */
+        void onResponseComplete(int errorCode) {
+            responseHandler.receiveResponseBodyComplete(errorCode);
+        }
+
+        /**
+         * Called from Native when the Http Request has a Body (Eg PUT/POST requests).
+         * @param out The Buffer to write the Body Bytes to.
+         * @return True if body stream is finished, false otherwise.
+         */
+        boolean sendOutgoingBody(ByteBuffer out) throws IOException {
+            boolean haveBody = request.getBody().isPresent();
+            if (!haveBody) {
+                return true;
+            }
+
+            byte[] tmp = new byte[out.remaining()];
+            InputStream bodyStream = request.getBody().get();
+            int amt_read = bodyStream.read(tmp); // TODO: This is a Blocking read. Ideally this would be non-blocking.
+
+            if (amt_read <= 0) {
+                return true;
+            }
+
+            out.put(tmp, 0, amt_read);
+            return false;
+
+        }
+    }
+
+    /**
+     * Interface used to receive the result of an async operation from CRT http
+     */
+    interface AsyncCallback {
```

Comment:
This is the same interface as the MQTT callback, should probably just move it to the CRT level and unify them.

Reply:
Fixed